src/read/read_ref.rs

```diff
@@ -0,0 +1,114 @@
+use alloc::vec::Vec;
+use core::convert::TryInto;
+use core::{mem, result};
+
+use crate::pod::{from_bytes, slice_from_bytes, Pod};
+
+type Result<T> = result::Result<T, ()>;
+
+/// A trait for reading references to `Pod` types from a block of data.
+///
+/// This allows parsers to handle both of these cases:
+/// - the block of data exists in memory, and it is desirable
+///   to use references to this block instead of copying it,
+/// - the block of data exists in storage, and it is desirable
+///   to read on demand to minimize I/O and memory usage.
+///
+/// The methods accept `self` by value because `Self` is expected to behave
+/// similar to a reference: it may be a reference with a lifetime of `'a`,
+/// or it may be a wrapper of a reference.
+///
+/// The `Clone` and `Copy` bounds are for convenience, and since `Self` is
+/// expected to be similar to a reference, these are easily satisfied.
+///
+/// Object file parsers typically use offsets to locate the structures
+/// in the block, and will most commonly use the `*_at` methods to
+/// read a structure at a known offset.
+///
+/// Occasionally file parsers will need to treat the block as a stream,
+/// and so convenience methods are provided that update an offset with
+/// the size that was read.
+//
+// An alternative would be for methods to accept `&mut self` and use a
+// `seek` method instead of the `offset` parameters, but this is less
+// convenient for implementers.
+pub trait ReadRef<'a>: Clone + Copy {
+    /// The total size of the block of data.
+    fn len(self) -> Result<u64>;
+
+    /// Get a reference to a `u8` slice at the given offset.
+    ///
+    /// Checks that the offset and size can be satisfied.
```

Comment:
Yeah, that's a problem I've been running into, too. I think you just can't free anything until the entire backing object goes away.
I'm not sure what the alternative is. If we want to keep the ability to return, for example, symbols with string names that refer back to the original data, with no additional copies for the "backed by memory slice" case compared to what we have today, then there's no choice but to keep that data around.
Would it make sense to have two different methods, `read_bytes_at` and `read_bytes_into`, with the latter doing a copy into an outparam? Then we can slowly transition the parsing code to use `read_bytes_into` for any data that doesn't need to be persistently retained. And `read_bytes_into` would also not have any alignment problems when reading POD data. But I'm not sure what impact this copy would have on parsing performance.

Reply:
> I'm not sure what the alternative is

Maybe I didn't spell it out clearly enough, but what I was trying to get at with the `Ref`/`RefMut` thing is that you can do is something like this:

```rust
// no more 'a in the trait
pub trait Read {
    fn read_at<'a>(&'a self, offset: usize, size: usize) -> Result<Ref<'a, [u8]>> {
        // ...
    }

    unsafe fn free<T>(&self, r: Ref<'_, T>) {
        // do the clean up here; maybe default empty function to get current leaky behavior by default

        // note that overlapping reads would need to also do ref counting internally or something
    }

    // ...
}

pub struct Ref<'a, T: ?Sized> {
    read: &'a dyn Read, // alternatively, make it a static type param
    inner: &'a T
}

impl<T: ?Sized> Deref for Ref<'_, T> {
    type Target = T;
    fn deref(&self) -> &T {
        // NB: this shortens the `'a` lifetime to `&self`'s lifetime! This way, borrowers can't hold
        // onto the slice forever, and we can reuse/free the buffer backing the data once `Ref` is
        // dropped.
        &self.inner
    }
}

impl<T: ?Sized> Drop for Ref<'_, T> {
    fn drop(&mut self) {
        unsafe {
            self.read.free(self.inner);
        }
    }
}

// Usage

let r: impl Read = ...;

{
    let data1: Ref<[u8]> = r.read_at(42, 1024)?;
    do_stuff(data);
    // data's drop impl will free the `Ref` at the end of this block
}

// data2 can reuse the space from `data1`
let data2 = r.read_at(0, 8)?;
```

Reply:
Thank you, that helped me a lot in understanding your suggestion.

I'll let Phillip comment on this. Personally my gut reaction is "help, smart pointers everywhere" but I'm not in a good position to evaluate the trade-off.

Reply:
The trade-off I'm trying to make here is to change nothing for the memory-mapped case, but hack in something that works good enough for the disk case. Ideally the disk case wouldn't be a hack, but I can't see how to do that without generic associated types (something like https://github.com/gimli-rs/object/issues/269#issuecomment-747808765, but I probably haven't thought through that example enough).

>     read: &'a dyn Read, // alternatively, make it a static type param

I think this means the compiler would have trouble optimizing this out for the `&'a [u8]` case. And if I understand correctly, making that a static type param would require generic associated types?

Reply:
I don't think it would require GATs since the only reason that field is needed is to call `free`, which can take a raw pointer+size or offset+size or something like that, rather than a `Ref`.

Reply:
(In my example, unlike the comment you link to, `Ref` is not an associated type, but always the same concrete type regardless which `Read` impl you have)

Reply:
I looked a bit into using `Ref` and it is complicated because we need to be able to borrow a subrange of a `Ref`. While it would be possible to overcome this, I don't think it is worth the complexity. (This problem would also occur with the GAT solution I linked to.) Additionally, replacing existing `&'data T` references with `Ref<T>` would be a large breaking change and a loss of ergonomics for the mmap use case.